actions/admin/contact.go

```diff
@@ -0,0 +1,159 @@
+package admin
+
+import (
+	"net/http"
+
+	"github.com/bitcoin-sv/spv-wallet/engine"
+	"github.com/bitcoin-sv/spv-wallet/mappings"
+	"github.com/bitcoin-sv/spv-wallet/models"
+	"github.com/gin-gonic/gin"
+)
+
+// contactsSearch will fetch a list of contacts filtered by Metadata and ContactFilters
+// Search for contacts filtering by metadata and ContactFilters godoc
+// @Summary		Search for contacts
+// @Description	Search for contacts
+// @Tags		Admin
+// @Produce		json
+// @Param		SearchRequestParameters body actions.SearchRequestParameters false "Supports targeted resource searches with filters for metadata and custom conditions, plus options for pagination and sorting to streamline data exploration and analysis"
+// @Success		200 {object} []models.Contact "List of contacts"
+// @Failure		400	"Bad request - Error while parsing SearchRequestParameters from request body"
+// @Failure 	500	"Internal server error - Error while searching for contacts"
+// @Router		/v1/admin/contact/search [post]
+// @Security	x-auth-xpub
+func (a *Action) contactsSearch(c *gin.Context) {
+	var reqParams SearchTransactions
+	if err := c.Bind(&reqParams); err != nil {
+		c.JSON(http.StatusBadRequest, err.Error())
+		return
+	}
+
+	// Record a new transaction (get the hex from parameters)a
+	contacts, err := a.Services.SpvWalletEngine.GetContacts(
+		c.Request.Context(),
+		reqParams.Metadata,
+		reqParams.Conditions.ToDbConditions(),
+		reqParams.QueryParams,
+	)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	contracts := make([]*models.Contact, 0)
+	for _, contact := range contacts {
+		contracts = append(contracts, mappings.MapToContactContract(contact))
+	}
+
+	c.JSON(http.StatusOK, contracts)
+}
+
+// contactsUpdate will update contact with the given id
+// Update contact FullName or Metadata godoc
+// @Summary		Update contact FullName or Metadata
+// @Description	Update contact FullName or Metadata
+// @Tags		Admin
+// @Produce		json
+// @Param		id path string false "Contact id"
+// @Param		UpdateContact body UpdateContact false "FullName and metadata to update"
+// @Success		200 {object} models.Contact "Updated contact"
+// @Failure		400	"Bad request - Error while parsing UpdateContact from request body or getting id from path"
+// @Failure 	500	"Internal server error - Error while updating contact"
+// @Router		/v1/admin/contact/{id} [patch]
+// @Security	x-auth-xpub
+func (a *Action) contactsUpdate(c *gin.Context) {
+	var reqParams UpdateContact
+	if err := c.Bind(&reqParams); err != nil {
+		c.JSON(http.StatusBadRequest, err.Error())
+		return
+	}
+
+	id := c.Param("id")
+	if id == "" {
+		c.JSON(http.StatusBadRequest, "id is required")
+		return
+	}
+
+	contact, err := a.Services.SpvWalletEngine.UpdateContact(
+		c.Request.Context(),
+		id,
+		reqParams.FullName,
+		&reqParams.Metadata,
+	)
+	if err != nil {
```

Comment:
Handle different errors:

```

	if err != nil {
		switch {
		case errors.Is(err, engine.ErrContactNotFound):
			c.JSON(http.StatusNotFound, err.Error())
		case errors.Is(err, engine.ErrContactIncorrectStatus):
			c.JSON(http.StatusUnprocessableEntity, err.Error())
		default:
			c.JSON(http.StatusInternalServerError, err.Error())
		}
		return
	}

Reply:
Done, prepared method to check it and it is used in all contact endpoints where different errors can be returned